01_Data/src/layers/sequelize/repository/TransactionEvent.ts

```diff
@@ -143,11 +155,74 @@ export class SequelizeTransactionEventRepository extends SequelizeRepository<Tra
     });
   }
 
+  private async calculateAndUpdateTotalKwh(transaction: Transaction): Promise<number> {
+    const totalKwh = this.getTotalKwh(transaction.meterValues ?? []);
+
+    await transaction.update(
+      {totalKwh}
+    );
+
+    return totalKwh;
+  }
+
+  private getTotalKwh(meterValues: MeterValueType[]): number {
+    const contexts: ReadingContextEnumType[] = [
+      ReadingContextEnumType.Transaction_Begin,
+      ReadingContextEnumType.Sample_Periodic,
+      ReadingContextEnumType.Transaction_End,
+    ];
+
+    let valuesMap = new Map();
+
+    meterValues
+      .filter(
+        (meterValue) =>
+          meterValue.sampledValue[0].context &&
+          contexts.indexOf(meterValue.sampledValue[0].context) !== -1,
+      )
+      .forEach((meterValue) => {
+        const sampledValues = meterValue.sampledValue as SampledValueType[];
+        const overallValue = sampledValues.find(
+          (sampledValue) =>
+            sampledValue.phase === undefined &&
+            sampledValue.measurand ===
+            MeasurandEnumType.Energy_Active_Import_Register,
+        );
+        if (
+          overallValue &&
+          overallValue.unitOfMeasure?.unit?.toUpperCase() === 'KWH'
+        ) {
+          valuesMap.set(Date.parse(meterValue.timestamp), overallValue.value);
+        } else if (
+          overallValue &&
+          overallValue.unitOfMeasure?.unit?.toUpperCase() === 'WH'
+        ) {
+          valuesMap.set(
+            Date.parse(meterValue.timestamp),
+            overallValue.value / 1000,
+          );
+        }
+      });
+
+    // sort the map based on timestamps
+    valuesMap = new Map(
+      [...valuesMap.entries()].sort((v1, v2) => v1[0] - v2[0]),
+    );
+    const sortedValues = Array.from(valuesMap.values());
+
+    let totalKwh = 0;
+    for (let i = 1; i < sortedValues.length; i++) {
```

Comment:
I would do it this way:
totalKwh = sortedValues[sortedValues.length-1] - sortedValues[0]
same answer, easier to understand and more performant

Reply:
Thanks! Will resolve all 🙌